File: android/src/main/java/com/svend/plugins/tcp/socket/TcpSocketPlugin.java
package com.svend.plugins.tcp.socket;

import android.Manifest;
import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.nio.charset.StandardCharsets;
import android.util.Base64;

/**
 * TcpSocketPlugin - Capacitor plugin that provides TCP Socket communication functionality
 * Supports connection, data sending, data receiving and disconnection operations
 */
@CapacitorPlugin(
    name = "TcpSocket",
    permissions = { @Permission(alias = "network", strings = { Manifest.permission.ACCESS_NETWORK_STATE }) }
)
public class TcpSocketPlugin extends Plugin {
    private static final String TAG = "TcpSocketPlugin";
    private List<Socket> clients = new ArrayList<>();
    
    // Supported data encoding types
    private enum DataEncoding {
        UTF8("utf8"),
        BASE64("base64"),
        HEX("hex");
        
        private final String value;
        
        DataEncoding(String value) {
            this.value = value;
        }
        
        public String getValue() {
            return value;
        }
        
        public static DataEncoding fromString(String text) {
            for (DataEncoding encoding : DataEncoding.values()) {
                if (encoding.value.equalsIgnoreCase(text)) {
                    return encoding;
                }
            }
            return UTF8; // Default to UTF8 if not found
        }
    }

    /**
     * Connect to TCP server
     * Parameters:
     * - ipAddress: Server IP address (required)
     * - port: Server port, default 9100
     * Returns:
     * - client: Client index, used for subsequent operations
     */
    @PluginMethod
    public void connect(PluginCall call) {
        String ipAddress = call.getString("ipAddress");

        if (ipAddress == null || ipAddress.isEmpty()) {
            call.reject("IP address is required");
            return;
        }
        Integer port = call.getInt("port", 9100);

        try {
            Socket socket = new Socket(ipAddress, port);
            clients.add(socket);
            
            JSObject ret = new JSObject();
            ret.put("client", clients.size() - 1);
            call.resolve(ret);
        } catch (IOException e) {
            Log.e(TAG, "Connection failed: " + e.getMessage());
            call.reject("Connection failed: " + e.getMessage());
        }
    }

    /**
     * Send data to server with specified encoding
     * Parameters:
     * - client: Client index
     * - data: Data to send
     * - encoding: Data encoding (utf8, base64, hex), default is utf8
     */
    @PluginMethod
    public void send(final PluginCall call) {
        final Integer clientIndex = call.getInt("client", -1);

        if (clientIndex == -1) {
            call.reject("Client not specified or invalid index");
            return;
        }
        
        if (clientIndex >= clients.size() || clientIndex < 0) {
            call.reject("Client index out of range");
            return;
        }
        
        final String data = call.getString("data");
        if (data == null || data.isEmpty()) {
            call.reject("No data provided");
            return;
        }
        
        final String encodingString = call.getString("encoding", "utf8");
        final DataEncoding encoding = DataEncoding.fromString(encodingString);

        Socket socket = clients.get(clientIndex);
        if (!socket.isConnected()) {
            closeSocketSafely(socket);
            call.reject("Socket not connected");
            return;
        }

        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                try {
                    final Socket socket = clients.get(clientIndex);
                    DataOutputStream bufferOut = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
                    
                    // Convert and send data according to the specified encoding
                    byte[] bytes;
                    switch (encoding) {
                        case BASE64:
                            bytes = Base64.decode(data, Base64.DEFAULT);
                            break;
                        case HEX:
                            bytes = hexStringToBytes(data);
                            break;
                        case UTF8:
                        default:
                            bytes = data.getBytes(StandardCharsets.UTF_8);
                            break;
                    }
                    
                    if (bytes != null) {
                        bufferOut.write(bytes);
                        bufferOut.flush();
                        call.resolve();
                    } else {
                        call.reject("Failed to decode data with encoding: " + encodingString);
                    }
                } catch (IOException e) {
                    Log.e(TAG, "Send failed: " + e.getMessage());
                    call.reject("Send failed: " + e.getMessage());
                }
            }
        };
        
        Thread thread = new Thread(runnable);
        thread.start();
    }

    /**
     * Read data from server
     * Parameters:
     * - client: Client index
     * - expectLen: Expected number of bytes to read, default 1024
     * - timeout: Timeout in seconds (Note: Android implementation doesn't support timeout yet)
     * - encoding: Preferred encoding for returned data (utf8, base64, hex), default is utf8
     * Returns:
     * - result: Data read
     * - encoding: The encoding used for the result
     */
    @PluginMethod
    public void read(final PluginCall call) {
        final Integer clientIndex = call.getInt("client", -1);
        final Integer length = call.getInt("expectLen", 1024);

        if (clientIndex == -1) {
            call.reject("Client not specified or invalid index");
            return;
        }
        
        if (clientIndex >= clients.size() || clientIndex < 0) {
            call.reject("Client index out of range");
            return;
        }

        Socket socket = clients.get(clientIndex);
        if (!socket.isConnected()) {
            closeSocketSafely(socket);
            call.reject("Socket not connected");
            return;
        }
        
        final String encodingString = call.getString("encoding", "utf8");
        final DataEncoding encoding = DataEncoding.fromString(encodingString);

        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                try {
                    final Socket socket = clients.get(clientIndex);
                    DataInputStream bufferIn = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
                    byte[] bytes = new byte[length];
                    int read = bufferIn.read(bytes, 0, length);

                    JSObject ret = new JSObject();
                    String result = "";
                    String actualEncoding = encoding.getValue();
                    
                    if (read > 0) {
                        byte[] actualData = new byte[read];
                        System.arraycopy(bytes, 0, actualData, 0, read);
                        
                        switch (encoding) {
                            case UTF8:
                                try {
                                    result = new String(actualData, StandardCharsets.UTF_8);
                                } catch (Exception e) {
                                    // If UTF-8 conversion fails, fall back to base64
                                    result = Base64.encodeToString(actualData, Base64.NO_WRAP);
                                    actualEncoding = DataEncoding.BASE64.getValue();
                                }
                                break;
                            case BASE64:
                                result = Base64.encodeToString(actualData, Base64.NO_WRAP);
                                break;
                            case HEX:
                                result = bytesToHexString(actualData);
                                break;
                        }
                    }
                    
                    ret.put("result", result);
                    ret.put("encoding", actualEncoding);
                    call.resolve(ret);
                } catch (IOException e) {
                    Log.e(TAG, "Read failed: " + e.getMessage());
                    call.reject("Read failed: " + e.getMessage());
                }
            }
        };

        Thread thread = new Thread(runnable);
        thread.start();
    }

    /**
     * Disconnect from server
     * Parameters:
     * - client: Client index
     * Returns:
     * - client: Disconnected client index
     */
    @PluginMethod
    public void disconnect(PluginCall call) {
        final Integer clientIndex = call.getInt("client", -1);
        if (clientIndex == -1) {
            call.reject("Client not specified or invalid index");
            return;
        }
        
        if (clients.isEmpty()) {
            call.reject("No active connections");
            return;
        }
        
        if (clientIndex >= clients.size() || clientIndex < 0) {
            call.reject("Client index out of range");
            return;
        }
        
        final Socket socket = clients.get(clientIndex);
        try {
            if (!socket.isConnected()) {
                socket.close();
                call.reject("Socket not connected");
                return;
            }
            socket.close();
        } catch (IOException e) {
            Log.e(TAG, "Disconnect failed: " + e.getMessage());
            call.reject("Disconnect failed: " + e.getMessage());
            return;
        }

        JSObject ret = new JSObject();
        ret.put("client", clientIndex);
        call.resolve(ret);
    }
    
    /**
     * Helper method to safely close a socket and handle any exceptions
     */
    private void closeSocketSafely(Socket socket) {
        if (socket != null) {
            try {
                socket.close();
            } catch (IOException e) {
                Log.e(TAG, "Error closing socket: " + e.getMessage());
            }
        }
    }
    
    /**
     * Clean up resources when plugin is destroyed
     */
    @Override
    protected void handleOnDestroy() {
        for (Socket socket : clients) {
            closeSocketSafely(socket);
        }
        clients.clear();
        super.handleOnDestroy();
    }
    
    /**
     * Convert hex string to bytes
     * @param hex Hex string to convert
     * @return Byte array or null if invalid hex string
     */
    private byte[] hexStringToBytes(String hex) {
        // Remove 0x prefix, spaces, newlines, etc.
        String cleanHex = hex.replace("0x", "")
                             .replace(" ", "")
                             .replace("\n", "");
                             
        if (cleanHex.length() % 2 != 0) {
            // Invalid hex string
            return null;
        }
        
        byte[] data = new byte[cleanHex.length() / 2];
        
        try {
            for (int i = 0; i < cleanHex.length(); i += 2) {
                data[i / 2] = (byte) Integer.parseInt(cleanHex.substring(i, i + 2), 16);
            }
            return data;
        } catch (NumberFormatException e) {
            Log.e(TAG, "Invalid hex string: " + e.getMessage());
            return null;
        }
    }
    
    /**
     * Convert bytes to hex string
     * @param bytes Byte array to convert
     * @return Hex string representation
     */
    private String bytesToHexString(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}

File: ios/Plugin/TcpSocketPlugin.swift
import Foundation
import Capacitor
import SwiftSocket

/**
 * TcpSocketPlugin - Capacitor plugin that provides TCP Socket communication functionality
 * Supports connection, data sending, data receiving and disconnection operations
 */
@objc(TcpSocketPlugin)
public class TcpSocketPlugin: CAPPlugin {
    /// Stores all connected TCP clients
    private var clients: [TCPClient] = []
    
    // Data encoding types
    private enum DataEncoding: String {
        case utf8 = "utf8"
        case base64 = "base64"
        case hex = "hex"
    }
    
    /**
     * Connect to TCP server
     * Parameters:
     * - ipAddress: Server IP address (required)
     * - port: Server port, default 9100
     * Returns:
     * - client: Client index, used for subsequent operations
     */
    @objc func connect(_ call: CAPPluginCall) {
        guard let ip = call.getString("ipAddress") else {
            call.reject("IP address is required")
            return
        }
        let port = Int32(call.getInt("port", 9100))

        let client = TCPClient(address: ip, port: port)
        switch client.connect(timeout: 10) {
        case .success:
            clients.append(client)
            call.resolve(["client": clients.count - 1])
        case .failure(let error):
            call.reject("Connection failed: \(error.localizedDescription)")
        }
    }
    
    /**
     * Send data to server with specified encoding
     * Parameters:
     * - client: Client index
     * - data: Data to send
     * - encoding: Data encoding (utf8, base64, hex), default is utf8
     */
    @objc func send(_ call: CAPPluginCall) {
        guard let clientIndex = call.getInt("client"), clientIndex >= 0 else {
            call.reject("Client not specified or invalid index")
            return
        }
        
        guard clients.indices.contains(clientIndex) else {
            call.reject("Client index out of range")
            return
        }
        
        let client = clients[clientIndex]
        guard let data = call.getString("data") else {
            call.reject("No data provided")
            return
        }
        
        let encodingString = call.getString("encoding", "utf8")
        guard let encoding = DataEncoding(rawValue: encodingString) else {
            call.reject("Unsupported encoding: \(encodingString)")
            return
        }
        
        var byteData: [UInt8]?
        
        // Convert the string to bytes based on the specified encoding
        switch encoding {
        case .utf8:
            byteData = [UInt8](data.utf8)
        case .base64:
            if let decodedData = Data(base64Encoded: data) {
                byteData = [UInt8](decodedData)
            }
        case .hex:
            byteData = hexStringToBytes(data)
        }
        
        guard let bytes = byteData else {
            call.reject("Failed to decode data with encoding: \(encodingString)")
            return
        }
        
        switch client.send(data: bytes) {
        case .success:
            call.resolve()
        case .failure(let error):
            call.reject("Send failed: \(error.localizedDescription)")
        }
    }
    
    /**
     * Read data from server
     * Parameters:
     * - client: Client index
     * - expectLen: Expected number of bytes to read, default 1024
     * - timeout: Timeout in seconds, default 10 seconds
     * - encoding: Preferred encoding for returned data (utf8, base64, hex), default is utf8
     * Returns:
     * - result: Data read
     * - encoding: The encoding used for the result
     */
    @objc func read(_ call: CAPPluginCall) {
        guard let clientIndex = call.getInt("client"), clientIndex >= 0 else {
            call.reject("Client not specified or invalid index")
            return
        }
        
        guard clients.indices.contains(clientIndex) else {
            call.reject("Client index out of range")
            return
        }
        
        let client = clients[clientIndex]
        let expectLen = call.getInt("expectLen", 1024)
        let timeout = call.getInt("timeout", 10)
        let encodingString = call.getString("encoding", "utf8")
        
        guard let encoding = DataEncoding(rawValue: encodingString) else {
            call.reject("Unsupported encoding: \(encodingString)")
            return
        }
        
        guard let response = client.read(expectLen, timeout: timeout) else {
            call.resolve(["result": "", "encoding": encoding.rawValue])
            return
        }
        
        let result: String
        
        switch encoding {
        case .utf8:
            // Try to convert to UTF-8 string, fall back to base64 if not possible
            if let utf8String = String(bytes: response, encoding: .utf8) {
                result = utf8String
            } else {
                // Fall back to base64 if UTF-8 conversion fails
                result = Data(response).base64EncodedString()
                call.resolve(["result": result, "encoding": DataEncoding.base64.rawValue])
                return
            }
        case .base64:
            result = Data(response).base64EncodedString()
        case .hex:
            result = bytesToHexString(response)
        }
        
        call.resolve(["result": result, "encoding": encoding.rawValue])
    }
    
    /**
     * Disconnect from server
     * Parameters:
     * - client: Client index
     * Returns:
     * - client: Disconnected client index
     */
    @objc func disconnect(_ call: CAPPluginCall) {
        guard let clientIndex = call.getInt("client"), clientIndex >= 0 else {
            call.reject("Client not specified or invalid index")
            return
        }
        
        guard clients.indices.contains(clientIndex) else {
            call.reject("Client index out of range")
            return
        }
        
        clients[clientIndex].close()
        call.resolve(["client": clientIndex])
    }
    
    /**
     * Convert hex string to bytes
     * - Parameter hex: Hex string (can have spaces and 0x prefix)
     * - Returns: Byte array
     */
    private func hexStringToBytes(_ hex: String) -> [UInt8]? {
        // Remove spaces, 0x prefixes, etc
        let cleanHex = hex.replacingOccurrences(of: "0x", with: "")
                          .replacingOccurrences(of: " ", with: "")
                          .replacingOccurrences(of: "\n", with: "")
        
        var bytes = [UInt8]()
        var i = cleanHex.startIndex
        
        while i < cleanHex.endIndex {
            let nextIndex = cleanHex.index(i, offsetBy: 2, limitedBy: cleanHex.endIndex) ?? cleanHex.endIndex
            let byteString = cleanHex[i..<nextIndex]
            
            if let byte = UInt8(byteString, radix: 16) {
                bytes.append(byte)
            } else {
                return nil
            }
            
            i = nextIndex
        }
        
        return bytes
    }
    
    /**
     * Convert bytes to hex string
     * - Parameter bytes: Byte array
     * - Returns: Hex string
     */
    private func bytesToHexString(_ bytes: [UInt8]) -> String {
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
}
